### PR TITLE
Combine metronome glue delegates into one

### PR DESCRIPTION
**Pull request by jason-hall**
Signed-off-by: Jason Hall <jasonhal@ca.ibm.com>

### CODE REVIEW

**Comment by dmitripivkine** [comment]
```
In file included from ../../gc_realtime/Scheduler.hpp:39:0,
                 from ../../gc_realtime/EnvironmentRealtime.hpp:30,
                 from ../../gc_realtime/ConfigurationRealtime.hpp:35,
                 from ../../gc_staccato/ConfigurationStaccato.hpp:35,
                 from ../../gc_staccato/ConfigurationStaccato.cpp:31:
../../gc_glue_java/MetronomeDelegate.hpp:77:31: error: ‘MM_EnvironmentRealtime’ has not been declared
  void incrementalCollectStart(MM_EnvironmentRealtime *env);
                               ^
../../gc_glue_java/MetronomeDelegate.hpp:78:26: error: ‘MM_EnvironmentRealtime’ has not been declared
  void incrementalCollect(MM_EnvironmentRealtime *env);
                          ^
In file included from ../../gc_realtime/Scheduler.hpp:39:0,
                 from ../../gc_realtime/EnvironmentRealtime.hpp:30,
                 from ../../gc_realtime/ConfigurationRealtime.hpp:35,
                 from ../../gc_staccato/ConfigurationStaccato.hpp:35,
                 from ../../gc_staccato/ConfigurationStaccato.cpp:31:
../../gc_glue_java/MetronomeDelegate.hpp:104:27: error: ‘MM_EnvironmentRealtime’ has not been declared
  void processDyingClasses(MM_EnvironmentRealtime *env, UDATA* classUnloadCountResult, UDATA* anonymousClassUnloadCount, UDATA* classLoaderUnloadCountResult, J9ClassLoader** classLoaderUnloadListResult);
                           ^
In file included from ../../gc_realtime/Scheduler.hpp:39:0,
                 from ../../gc_realtime/EnvironmentRealtime.hpp:30,
                 from ../../gc_realtime/ConfigurationRealtime.hpp:35,
                 from ../../gc_staccato/ConfigurationStaccato.hpp:35,
                 from ../../gc_staccato/ConfigurationStaccato.cpp:31:
../../gc_glue_java/MetronomeDelegate.hpp:117:11: error: expected ‘;’ at end of member declaration
  J9Class *addDyingClassesToList(MM_EnvironmentRealtime *env, J9ClassLoader * classLoader, bool setAll, J9Class *classUnloadListStart, UDATA *classUnloadCountResult);
           ^
../../gc_glue_java/MetronomeDelegate.hpp:117:56: error: expected ‘)’ before ‘*’ token
  J9Class *addDyingClassesToList(MM_EnvironmentRealtime *env, J9ClassLoader * classLoader, bool setAll, J9Class *classUnloadListStart, UDATA *classUnloadCountResult);
                                                        ^
../../gc_glue_java/MetronomeDelegate.hpp:120:31: error: ‘MM_EnvironmentRealtime’ has not been declared
  void yieldFromClassUnloading(MM_EnvironmentRealtime *env);
                               ^
../../gc_glue_java/MetronomeDelegate.hpp:121:30: error: ‘MM_EnvironmentRealtime’ has not been declared
  void lockClassUnloadMonitor(MM_EnvironmentRealtime *env);
                              ^
../../gc_glue_java/MetronomeDelegate.hpp:122:32: error: ‘MM_EnvironmentRealtime’ has not been declared
  void unlockClassUnloadMonitor(MM_EnvironmentRealtime *env);
                                ^
../../gc_glue_java/MetronomeDelegate.hpp:136:22: error: ‘MM_EnvironmentRealtime’ has not been declared
  bool doClassTracing(MM_EnvironmentRealtime* env);
                      ^
../../gc_glue_java/MetronomeDelegate.hpp:138:25: error: ‘MM_EnvironmentRealtime’ has not been declared
  virtual bool doTracing(MM_EnvironmentRealtime* env);
                         ^
gc_staccato/CMakeFiles/j9staccato.dir/build.make:62: recipe for target 'gc_staccato/CMakeFiles/j9staccato.dir/ConfigurationStaccato.cpp.o' failed
make[6]: Leaving directory '/home/travis/build/eclipse/openj9-openjdk-jdk11/build/linux-x86_64-normal-server-release/vm/build'
make[6]: *** [gc_staccato/CMakeFiles/j9staccato.dir/ConfigurationStaccato.cpp.o] Error 1
CMakeFiles/Makefile2:3499: recipe for target 'gc_staccato/CMakeFiles/j9staccato.dir/all' failed
make[5]: *** [gc_staccato/CMakeFiles/j9staccato.dir/all] Error 2
```

**Comment by jason-hall** [comment]
Yes, thanks, I can see it.

**Comment by dmitripivkine** [comment]
There in number of unchanged files in change set but should be excluded:

runtime/gc_realtime/AllocationContextRealtime.cpp
runtime/gc_realtime/AllocationContextRealtime.hpp
runtime/gc_realtime/ConfigurationRealtime.cpp
runtime/gc_realtime/ConfigurationRealtime.hpp
runtime/gc_realtime/HeapRegionDescriptorRealtime.cpp
runtime/gc_realtime/HeapRegionDescriptorRealtime.hpp

**Comment by jason-hall** [comment]
Previous change mistakenly changed the file permissions to add executable. These aren't unchanged files, its restoring the file permissions. I don't think its possible to add unchanged files.

**Comment by dmitripivkine** [comment]
> Previous change mistakenly changed the file permissions to add executable. These aren't unchanged files, its restoring the file permissions. I don't think its possible to add unchanged files.

ok. Yes, it is possible to add unchanged file. One of scenarios if you modify file in first commit but undo changes in second. Unchanged file would be part of change set until removed manually.

**Comment by dmitripivkine** [comment]
Jenkins compile win

**Comment by dmitripivkine** [comment]
Jenkins test sanity xLinux jdk8

**Comment by dmitripivkine** [comment]
```
18:18:25  j9realtime.lib(MetronomeAlarmThread.obj) : error LNK2019: unresolved external symbol "public: static int __cdecl MM_MetronomeDelegate::metronomeAlarmThreadWrapper(void *)" (?metronomeAlarmThreadWrapper@MM_MetronomeDelegate@@SAHPEAX@Z) referenced in function "private: static int __cdecl MM_MetronomeAlarmThread::metronomeAlarmThreadWrapper(void *)" (?metronomeAlarmThreadWrapper@MM_MetronomeAlarmThread@@CAHPEAX@Z)
18:18:25  ../j9gc29.dll : fatal error LNK1120: 1 unresolved externals
18:18:25  make[3]: *** [../makelib/targets.mk:472: ../j9gc29.dll] Error 96
18:18:25  make[2]: *** [makefile:1015: j9gc] Error 2
```

File: runtime/gc_glue_java/MetronomeDelegate.cpp [hunk 1]
```diff
@@ -20,34 +20,102 @@
...
+		}
+		MM_JNICriticalRegion::reacquireAccess((J9VMThread *)env->getOmrVMThread()->_language_vmthread, accessMask);
+	}
+}
+
+/**
+ * C entrypoint for the newly created alarm thread.
+ */
+int J9THREAD_PROC
+MM_MetronomeAlarmThreadDelegate::metronomeAlarmThreadWrapper(void* userData)
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/MetronomeDelegate.cpp]: should `MM_MetronomeAlarmThreadDelegate` be changed to `MM_MetronomeDelegate`

File: runtime/gc_glue_java/MetronomeDelegate.cpp [hunk 2]
```diff
@@ -20,34 +20,102 @@
...
+{
+	MM_MetronomeAlarmThread *alarmThread = (MM_MetronomeAlarmThread *)userData;
+	J9JavaVM *javaVM = (J9JavaVM *)alarmThread->getScheduler()->_extensions->getOmrVM()->_language_vm;
+	PORT_ACCESS_FROM_JAVAVM(javaVM);
+	uintptr_t rc;
+
+	j9sig_protect(MM_MetronomeAlarmThreadDelegate::signalProtectedFunction, (void*)userData,
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/MetronomeDelegate.cpp]: should `MM_MetronomeAlarmThreadDelegate` be changed to `MM_MetronomeDelegate `

File: runtime/gc_glue_java/MetronomeDelegate.hpp [hunk 3]
```diff
@@ -122,6 +127,16 @@ class MM_RealtimeGCDelegate : public MM_BaseNonVirtual
...
+	MM_RealtimeAccessBarrier* allocateAccessBarrier(MM_EnvironmentBase *env);
+	void enableDoubleBarrier(MM_EnvironmentBase* env);
+	void disableDoubleBarrierOnThread(MM_EnvironmentBase* env, OMR_VMThread* vmThread);
+	void disableDoubleBarrier(MM_EnvironmentBase* env);
+
+	/* New methods */
+#if defined(J9VM_GC_DYNAMIC_CLASS_UNLOADING)
+	bool doClassTracing(MM_EnvironmentRealtime* env);
+#endif /* J9VM_GC_DYNAMIC_CLASS_UNLOADING */
+	virtual bool doTracing(MM_EnvironmentRealtime* env);
```
Comment (dmitripivkine) [inline runtime/gc_glue_java/MetronomeDelegate.hpp]: Why this method is `virtual` ?

File: runtime/gc_realtime/MemorySubSpaceMetronome.cpp [hunk 4]
```diff
@@ -86,10 +86,10 @@ MM_MemorySubSpaceMetronome::collectOnOOM(MM_EnvironmentBase *env, MM_GCCode gcCo
 	if (sched->isInitialized()) {
 		sched->startGC(envRealtime);
 		sched->setGCCode(gcCode);
-		sched->continueGC(envRealtime, OUT_OF_MEMORY_TRIGGER, allocDescription->getBytesRequested(), env->getOmrVMThread(), true);
+		sched->continueGC(envRealtime, OUT_OF_MEMORY_TRIGGER, allocDescription->getBytesRequested(), (OMR_VMThread *)env->getOmrVMThread(), true);
```
Comment (dmitripivkine) [inline runtime/gc_realtime/MemorySubSpaceMetronome.cpp]: `env->getOmrVMThread()` has type `(OMR_VMThread *)`... no extra cast necessary

File: runtime/gc_realtime/MemorySubSpaceMetronome.cpp [hunk 5]
```diff
@@ -153,9 +153,9 @@ MM_MemorySubSpaceMetronome::systemGarbageCollect(MM_EnvironmentBase *env, U_32 g
 		sched->startGC(envRealtime);
 		sched->setGCCode(MM_GCCode(gcCode));
 		/* if we were triggered by rasdump, then the caller has already acquired exclusive VM access */
-		sched->continueGC(envRealtime, SYSTEM_GC_TRIGGER, 0, env->getOmrVMThread(), J9MMCONSTANT_EXPLICIT_GC_RASDUMP_COMPACT != gcCode);
+		sched->continueGC(envRealtime, SYSTEM_GC_TRIGGER, 0, (OMR_VMThread *)envRealtime->getOmrVMThread(), J9MMCONSTANT_EXPLICIT_GC_RASDUMP_COMPACT != gcCode);
```
Comment (dmitripivkine) [inline runtime/gc_realtime/MemorySubSpaceMetronome.cpp]: why this change necessary?

Comment (jason-hall) [reply runtime/gc_realtime/MemorySubSpaceMetronome.cpp]: Its clearly not necessary. Its the result of repeatedly rebasing and redoing my changes dozens of times.

**Comment by dmitripivkine** [comment]
Jenkins compile win

**Comment by dmitripivkine** [comment]
Jenkins test sanity xLinux jdk8

**Comment by jason-hall** [comment]
Looks like I broke the includes with a change intended for another branch

**Comment by dmitripivkine** [comment]
Jenkins compile win

**Comment by dmitripivkine** [comment]
Jenkins test sanity xLinux jdk8

**Comment by dmitripivkine** [comment]
Jenkins compile win

**Comment by dmitripivkine** [comment]
```
16:32:47  not ok 90 - J9vmTest_2
16:32:47    ---
16:32:47      output:
16:32:47        |
16:32:47          ===============================================
16:32:47          Running test J9vmTest_2 ...
16:32:47          ===============================================
16:32:47          J9vmTest_2 Start Time: Fri Mar 29 17:00:51 2019 Epoch Time (ms): 1553889651187
16:32:47          variation: Mode351
16:32:47          JVM_OPTIONS: -Xcompressedrefs -Xgcpolicy:metronome -Xcompressedrefs 
16:32:47          Unhandled exception
16:32:47          Type=Segmentation error vmState=0x00000000
16:32:47          J9Generic_Signal_Number=00000004 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
16:32:47          Handler1=00007FDED1F37BB0 Handler2=00007FDED1321330 InaccessibleAddress=0000000000000000
16:32:47          RDI=0000000000000000 RSI=0000000001B49488 RAX=00007FDECC1FD5B0 RBX=00007FDECC16AD78
16:32:47          RCX=0000000000000000 RDX=0000000000000001 R8=00007FDECC1FD858 R9=00007FDECC1FD850
16:32:47          R10=00007FDECC031B78 R11=00007FDECC031B80 R12=0000000001B48700 R13=00007FDECC015B20
16:32:47          R14=0000000001B48700 R15=0000000000000000
16:32:47          RIP=00007FDECBC5E0B4 GS=0000 FS=0000 RSP=00007FDED2E95360
16:32:47          EFlags=0000000000010246 CS=0033 RBP=00007FDECC02C000 ERR=0000000000000004
16:32:47          TRAPNO=000000000000000E OLDMASK=0000000000000000 CR2=0000000000000000
16:32:47          xmm0 0000000000000100 (f: 256.000000, d: 1.264808e-321)
16:32:47          xmm1 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm2 6974736f6e676169 (f: 1852268928.000000, d: 9.783854e+199)
16:32:47          xmm3 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm4 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm5 5b5b5b5b5b5b5b5b (f: 1532713856.000000, d: 1.213625e+132)
16:32:47          xmm6 2020202020202020 (f: 538976256.000000, d: 6.013470e-154)
16:32:47          xmm7 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm8 68542f656b6f766e (f: 1802466944.000000, d: 3.683740e+194)
16:32:47          xmm9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm10 00000000ffff0000 (f: 4294901760.000000, d: 2.121963e-314)
16:32:47          xmm11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm12 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm13 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          xmm15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
16:32:47          Module=/home/jenkins/workspace/PullRequest-Sanity-JDK8-linux_x86-64_cmprssptrs-OpenJ9/build/linux-x86_64-normal-server-release/images/j2sdk-image/jre/lib/amd64/compressedrefs/libj9gc29.so
16:32:47          Module_base_address=00007FDECBB24000
16:32:47          Target=2_90_20190329_795 (Linux 4.4.0-140-generic)
16:32:47          CPU=amd64 (4 logical CPUs) (0x5e2f18000 RAM)
16:32:47          ----------- Stack Backtrace -----------
16:32:47          (0x00007FDECBC5E0B4 [libj9gc29.so+0x13a0b4])
16:32:47          (0x00007FDECBB8A81E [libj9gc29.so+0x6681e])
16:32:47          (0x00007FDECBC5E2C3 [libj9gc29.so+0x13a2c3])
16:32:47          (0x00007FDECBC9AF06 [libj9gc29.so+0x176f06])
16:32:47          (0x00007FDECBB6BFF9 [libj9gc29.so+0x47ff9])
16:32:47          (0x00007FDED1F718B7 [libj9vm29.so+0xd98b7])
16:32:47          (0x00007FDED1F430E8 [libj9vm29.so+0xab0e8])
16:32:47          (0x00007FDED13222D0 [libj9prt29.so+0x202d0])
16:32:47          (0x00007FDED1F4378E [libj9vm29.so+0xab78e])
16:32:47          (0x00007FDED1F52AEA [libj9vm29.so+0xbaaea])
16:32:47          (0x00007FDED13222D0 [libj9prt29.so+0x202d0])
16:32:47          (0x00007FDED1F4C59D [libj9vm29.so+0xb459d])
16:32:47          J9_CreateJavaVM+0x85 (0x00007FDED1F432D5 [libj9vm29.so+0xab2d5])
16:32:47          JNI_CreateJavaVM+0x7a4 (0x00007FDED245A864 [libjvm.so+0x18864])
16:32:47          JNI_CreateJavaVM+0xa26 (0x00007FDED2EA2806 [libjvm.so+0xa806])
16:32:47          (0x00007FDED38A01A4 [libjli.so+0x81a4])
16:32:47          (0x00007FDED30B46BA [libpthread.so.0+0x76ba])
16:32:47          clone+0x6d (0x00007FDED35D541D [libc.so.6+0x10741d])
16:32:47          ---------------------------------------
16:32:47          JVMDUMP039I Processing dump event "gpf", detail "" at 2019/03/29 17:00:51 - please wait.
16:32:47          JVMDUMP032I JVM requested System dump using '/home/jenkins/workspace/PullRequest-Sanity-JDK8-linux_x86-64_cmprssptrs-OpenJ9/openj9/test/TestConfig/test_output_15538879727612/J9vmTest_2/core.20190329.170051.30083.0001.dmp' in response to an event
16:32:47          JVMPORT030W /proc/sys/kernel/core_pattern setting "|/usr/share/apport/apport %p %s %c %d %P" specifies that the core dump is to be piped to an external program.  Attempting to rename either core or core.30086.
```

**Comment by dmitripivkine** [comment]
Jenkins compile win

**Comment by dmitripivkine** [comment]
Jenkins test sanity xLinux jdk8

**Comment by dmitripivkine** [comment]
Jenkins test sanity AIX jdk11
